[NFC][FMV][AArch64] Tidy up codegen tests.

labrinea opened the pull request:
Removes attr-target-version.c which doesn't have a clear purpose.
Introduces AArch64/fmv-detection.c to check detection bitmasks.
Adds coverage in AArch64/fmv-resolver-emission.c

labrinea committed:
[NFC][FMV][AArch64] Tidy up codegen tests.

Removes attr-target-version.c which doesn't have a clear purpose.
Introduces AArch64/fmv-detection.c to check detection bitmasks.
Adds coverage in AArch64/fmv-resolver-emission.c

tmatheson-arm commented on clang/test/CodeGen/attr-target-version.c:
```diff
@@ -1,1145 +0,0 @@
-__attribute__((target_version("aes"))) int unused_with_default_decl(void) { return 0; }
-__attribute__((target_version("default"))) int unused_with_default_decl(void);
-
-// This should generate two target versions and the resolver.
-__attribute__((target_version("sve"))) int unused_with_default_def(void) { return 0; }
-__attribute__((target_version("default"))) int unused_with_default_def(void) { return 1; }
-
-// This should also generate two target versions and the resolver.
-__attribute__((target_version("fp16"))) int unused_with_implicit_default_def(void) { return 0; }
-int unused_with_implicit_default_def(void) { return 1; }
-
-// This should also generate two target versions and the resolver.
-int unused_with_implicit_forward_default_def(void) { return 0; }
-__attribute__((target_version("lse"))) int unused_with_implicit_forward_default_def(void) { return 1; }
-
-// This should generate a target version despite the default not being declared.
-__attribute__((target_version("rdm"))) int unused_without_default(void) { return 0; }
-
-// These shouldn't generate anything.
-int unused_version_declarations(void);
-__attribute__((target_version("jscvt"))) int unused_version_declarations(void);
-__attribute__((target_version("rdma"))) int unused_version_declarations(void);
-
-// These should generate the default (mangled) version and the resolver.
-int default_def_with_version_decls(void) { return 0; }
-__attribute__((target_version("jscvt"))) int default_def_with_version_decls(void);
-__attribute__((target_version("rdma"))) int default_def_with_version_decls(void);
-
-// The following is guarded because in NOFMV we get errors for calling undeclared functions.
-#ifdef __HAVE_FUNCTION_MULTI_VERSIONING
-// This should generate a default declaration, two target versions but no resolver.
-__attribute__((target_version("jscvt"))) int used_def_without_default_decl(void) { return 1; }
-__attribute__((target_version("rdma"))) int used_def_without_default_decl(void) { return 2; }
-
-// This should generate a default declaration but no resolver.
-__attribute__((target_version("jscvt"))) int used_decl_without_default_decl(void);
-__attribute__((target_version("rdma"))) int used_decl_without_default_decl(void);
-
-int caller(void) { return used_def_without_default_decl() + used_decl_without_default_decl(); }
-#endif
```
I'm a bit wary of blanket-removing tests. These ones look useful, testing interaction of the attribute with other language features, and they have comments describing the point of them.

labrinea replied:
I have added those tests. We have coverage for these in clang/test/CodeGen/AArch64/fmv-resolver-emission.c

labrinea replied:
We also have coverage in clang/test/CodeGen/AArch64/fmv-mix-explicit-implicit-default.c

labrinea commented on clang/test/CodeGen/attr-target-version.c:
```diff
@@ -1,1145 +0,0 @@
-// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -emit-llvm -o - %s | FileCheck %s
-// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature -fmv -emit-llvm -o - %s | FileCheck %s -check-prefix=CHECK-NOFMV
-
-int __attribute__((target_version("rng+flagm+fp16fml"))) fmv(void) { return 1; }
```
We have semantic and name mangling tests in place. I may add a test to show the bitmask construction works. Also I may be adding a test to show that features coming from the command line are not exempt from detection.

labrinea commented on clang/test/CodeGen/AArch64/fmv-detection.c:
```diff
@@ -0,0 +1,790 @@
+__attribute__((target_version("default"))) int fmv(void);
+
+int caller() {
+  return fmv();
```
I am not seeing much value on this. We only want to see the ifunc symbol being referenced upon call.

labrinea committed:
Changes in clang/test/CodeGen/AArch64/fmv-detection.c
 - Convert run line to clang_cc1.
 - Check that cmdline features are not exempt from detection.
 - Add a version which uses more than one features.
Autogenerate clang/test/CodeGen/AArch64/fmv-resolver-emission.c

tommat01 reviewed (approved):
I can't check 100% we are not missing anything important, but if you are confident then go for it.

labrinea commented:
Cool, thanks Tomas

labrinea committed:
Merge branch 'main' into fmv-cleanup-codegen-tests
